### PR TITLE
Cm.2

### PR DESCRIPTION
**Pull request by redheadedrod**
Need to merge  CM.1 first. More fixes

### CODE REVIEW

File: envi/archs/arm/const.py [hunk 1]
```diff
@@ -333,10 +340,10 @@ def instrenc(encoding, index):
 INS_BCC     = instrenc(IENC_BRANCH, 2)
 INS_BX      = instrenc(IENC_MISC, 3)
 INS_BXJ     = instrenc(IENC_MISC, 5)
-INS_BLX     = IENC_UNCOND_BLX
+#INS_BLX     = IENC_UNCOND_BLX
```
Comment (atlas0fd00m) [inline envi/archs/arm/const.py]: nope.  please leave this as INS_BLX.
i'd like to move to using INS_<mnem> as the core of everything during decoding.

File: envi/archs/arm/disasm.py [hunk 2]
```diff
@@ -406,7 +406,7 @@ def p_misc1(opval, va): #
         )
     elif opval & 0x0ff000f0 == 0x01200030:
         #opcode = (IENC_MISC << 16) + 6
-        opcode = INS_BLX
+        opcode = IENC_UNCOND_BLX
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: this as well :)

File: envi/archs/arm/disasm.py [hunk 3]
```diff
@@ -430,10 +430,8 @@ def p_misc1(opval, va): #
         mnem = 'bkpt'
         immed = ((opval>>4)&0xfff0) + (opval&0xf)
         olist = ( ArmImmOper(immed), )
-    #dbg - not sure where to put this
     elif opval & 0xfff00f0 == 0x32000f0:
-        #need opcode
-        opcode = (IENC_MISC << 16) + 8
+        opcode = (IENC_MISC << 16) + 8 #FIXME - needs opcode
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: INS_DBG, please create it :)

File: envi/archs/arm/disasm.py [hunk 4]
```diff
@@ -766,19 +763,18 @@ def p_dp_movw(opval, va):
     iflags = 0
     imm =  ((opval >>4) &0xf000) + (opval & 0xfff)
     Rd = (opval >> 12) & 0xf
-    opcode = INS_MOV
+    opcode = INS_MOV #FIXME
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: what's wrong here?

File: envi/archs/arm/disasm.py [hunk 5]
```diff
@@ -766,19 +763,18 @@ def p_dp_movw(opval, va):
...
     )
     return(opcode, "movw", olist, iflags)
 
 def p_dp_movt(opval, va):
-    #fix opcode
     iflags = 0
     imm =  ((opval >>4) &0xf000) + (opval & 0xfff)
     Rd = (opval >> 12) & 0xf
-    opcode = INS_MOV
+    opcode = INS_MOV #FIXME
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: and here?

File: envi/archs/arm/disasm.py [hunk 6]
```diff
@@ -1482,26 +1509,25 @@ def p_uncond(opval, va):
             #clrex
             mnem = "clrex"
             olist =()
-            opcode = INS_CLREX
+            opcode = IENC_UNCOND_CLREX
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: :) i don't think this is helpful.

IENC_ means "Instruction Encoding Scheme" (which made sense for the initial ARM breakdown, but less so for the Thumb instructions).
INS_ means "Instruction".

File: envi/archs/arm/disasm.py [hunk 7]
```diff
@@ -1482,26 +1509,25 @@ def p_uncond(opval, va):
...
             return (opcode, mnem, olist, 0)
         elif (opval & 0xff000e0) == 0x5700040:
             #dmb/dsb
             option = opval & 0xf
             if (opval & 0x10 )== 0x10:
                 mnem = 'dmb'
-                opcode = INS_DMB
+                opcode = IENC_UNCOND_DMB
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: newp.

File: envi/archs/arm/disasm.py [hunk 8]
```diff
@@ -1482,26 +1509,25 @@ def p_uncond(opval, va):
...
             else:
                 mnem = 'dsb'
-                opcode = INS_DSB
+                opcode = IENC_UNCOND_DSB
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: and here....
sorry for the confusion.  this is my fault.

File: envi/archs/arm/disasm.py [hunk 9]
```diff
@@ -1482,26 +1509,25 @@ def p_uncond(opval, va):
...
             olist = (ArmBarrierOption(option),)
-            
             return (opcode, mnem, olist, 0)
         elif (opval & 0xff000f0) == 0x5700060:
             #isb
             option = opval & 0xf
             mnem = 'isb'
             olist = (ArmBarrierOption(option),)
-            opcode = INS_ISB
+            opcode = IENC_UNCOND_ISB
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: newp

File: envi/archs/arm/disasm.py [hunk 10]
```diff
@@ -1550,7 +1576,7 @@ def p_uncond(opval, va):
                 ArmPcOffsetOper(imm_offset, va),
             )
             
-            opcode = INS_BLX           #should this be IENC_UNCOND_BLX?
+            opcode = IENC_UNCOND_BLX
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: newp

File: envi/archs/arm/disasm.py [hunk 11]
```diff
@@ -1829,7 +1855,7 @@ def getBranches(self, emu=None):
         if self.prefixes != COND_AL:
             flags |= envi.BR_COND
 
-        if self.opcode in ( INS_B, INS_BX, INS_BL, INS_BLX, INS_BCC ):
+        if self.opcode in ( INS_B, INS_BX, INS_BL, IENC_UNCOND_BLX, INS_BCC ):
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: newp.

File: envi/archs/arm/disasm.py [hunk 12]
```diff
@@ -1838,7 +1864,7 @@ def getBranches(self, emu=None):
                 # probably a branch to a register.  just return.
                 return ret
 
-            if self.opcode in (INS_BLX, INS_BX):
+            if self.opcode in (IENC_UNCOND_BLX, INS_BX):
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: newp.

File: envi/archs/arm/disasm.py [hunk 13]
```diff
@@ -2562,9 +2588,9 @@ def repr(self, op):
             addr = self.getOperAddr(op)    # only works without an emulator because we've already verified base_reg is PC
             tname = "[#0x%x]" % addr
             # FIXME: is there any chance of us doing indexing on PC?!?
-            # ldcl literal trips this in some cases - leaving for now - to check for other instances
-            if idxing != 0x10:
-                print "OMJ! indexing on the program counter!"
+            # ldcl literal trips this in some cases
+            #if idxing != 0x10:
+            #    print "OMJ! indexing on the program counter!"
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: i need to investigate these and probably just remove them.  we seem to index on PC more often than we originally thought possible.

File: envi/archs/arm/disasm.py (hunk 4 above)
Comment (redheadedrod) [inline envi/archs/arm/disasm.py]: Nothing, comment is removed

File: envi/archs/arm/disasm.py (hunk 5 above)
Comment (redheadedrod) [inline envi/archs/arm/disasm.py]: Comment Removed

File: envi/archs/arm/disasm.py (hunk 13 above)
Comment (redheadedrod) [inline envi/archs/arm/disasm.py]: Removed in later PR.

**Comment by atlas0fd00m** [comment]
I believe you need to update atlas_arm_fixes, and merge it into this branch.

**Pull request merged**